### ISSUE TITLE
GaussianSplatting densification (split/prune during training) not exposed through the facade

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1835]
## Problem

The 3D Gaussian Splatting paper's key training-time mechanism is **densification** — periodically SPLIT Gaussians whose accumulated gradient magnitude exceeds a threshold (they cover too much space) and PRUNE Gaussians whose opacity has collapsed to near-zero. This grows the initial cloud (~100k random Gaussians) into millions of specialized ones over the training run, and is what gives real GS its paper-quality output.

`GaussianSplattingOptions` exposes fields like `DensificationInterval`, `SplitPositionJitter`, `SplitScaleFactor`, `SplitOpacityFactor` — the config knobs exist. But nothing in the facade path (`ConfigureModel` → `BuildAsync`) invokes densification during training. The Gaussian count stays fixed at whatever the constructor produced.

Combined with #1833 (facade optimizer settings silently ignored for GS), this caps GS quality at "random-init cloud with tiny gradient updates" — which is what the Week 11 Session B demo demonstrates visually (a few blurry blobs, unchanged across camera angles, regardless of training epochs).

## Fix direction (agreed in design discussion)

**Model-internal densification, following DDPM's specialization pattern.**

DDPM handles its specialization (noise-prediction training, timestep sampling) entirely inside its inherited `NeuralNetworkBase.Train(input, target)` path — callers see a uniform `Train(x, y)` API; the specialized behavior is internal. No callback infrastructure was needed. GS densification is the same shape of problem — an interleaved-during-training operation that mutates the model — so it applies the same pattern.

Concretely: densification lives inside `GaussianSplatting.TrainOnImageBatch` (per #1834's `IImageTrainable<T>`). `GaussianSplattingOptions.DensificationInterval` and friends drive the schedule; the model decides when to densify based on its options. Facade calls `TrainOnImageBatch` per iteration; the model runs its gradient step + interleaves densification when the schedule fires. No new callback interface, no divergence from how DDPM already works.

## Why this exceeds industry standard

- Reference GS implementations (gsplat, nerfstudio, INRIA's official 3DGS repo) hardcode densification inside their training scripts. Users configure via YAML sections but the schedule is baked in — one code path per fork.
- AiDotNet with model-internal densification: densification is a first-class configurable property of `GaussianSplattingOptions` — **same options object exposes densification + hyperparameters + rendering** all in one place. Reference impls fragment these across YAML sections (model config vs. training config vs. optimizer config).
- No new callback infrastructure to teach; no divergence from how DDPM (and future specialized models) work. Uniform mental model.

## Beyond-industry excellence goals

Four goals, all agreed to during design discussion:

### 1. All densification config in one strongly-typed options object (already partially there)
`GaussianSplattingOptions` already exposes `DensificationInterval`, `SplitPositionJitter`, `SplitScaleFactor`, `SplitOpacityFactor`, `SplitOpacityMax`. Extend to cover **every** industry-standard knob:
- `DensificationStartIteration` — when to begin densifying (skip first N iters while cloud stabilizes)
- `DensificationEndIteration` — when to stop (near end of training, freeze the cloud)
- `GradientNormThreshold` — the actual split trigger (paper: `τ_pos = 0.0002`)
- `OpacityPruneThreshold` — opacity below which a Gaussian is culled (paper: `0.005`)
- `MaxGaussianCount` — hard ceiling to prevent OOM
- `GradientAccumulationWindow` — how many iterations of gradient norm to average for the split decision
All nullable, all industry-standard defaults (from the 3DGS paper), user can override any.

### 2. Adaptive densification schedule (data-driven, not fixed interval)
Reference impls use a fixed `every N iterations` schedule. AiDotNet exceeds by exposing an `AdaptiveSchedule` option (default off, matches industry when off) that triggers densification when observed signals warrant it:
- Gradient norm variance drops (cloud has stabilized enough that split noise won't destabilize)
- Loss plateaus (cloud can't improve without more capacity)
Adaptive schedule is fully customizable via a strategy interface; industry-standard fixed-interval schedule remains the default.

### 3. Post-training densification pruning + compression pass
After training, run a compression pass:
- Prune all Gaussians below `OpacityPruneThreshold`
- Merge nearby Gaussians whose bounding ellipses overlap by > threshold
- Quantize spherical harmonics (float32 → int8 with a per-Gaussian scale)

Reference impls do this manually as post-processing scripts. AiDotNet exposes it as a first-class option: `CompressOnBuildComplete = true` (default `false` — industry-standard: user runs their own post-processing). When enabled, `BuildAsync` returns with the compressed cloud. Fully customizable via `CompressionOptions` (per-technique on/off).

### 4. Per-attribute learning rate schedules for split children
When a Gaussian is split, the two children inherit modified hyperparameters:
- Position LR **decreases** (they're now local — small movements matter more, big jumps destabilize)
- Scale LR **increases** (they're smaller — need more updates per step to converge)
- Opacity LR unchanged
Paper implements this manually. AiDotNet exposes via `SplitChildLearningRateScales` — nullable per-attribute multipliers, industry-standard defaults from the paper, user can override.

## Wiring surface

- `ConfigureModel` (existing) — no change; `GaussianSplattingOptions` extended.
- `ConfigureOptimizer` (existing) — feeds into the per-attribute schedules (per #1833's hyperparameter routing).
- No new `Configure*` methods.

## Related

- #1826 / #1829 — original facade unblock chain
- #1833 — GS optimizer hyperparameter routing (densification per-attribute LR schedules use this)
- #1834 — image-space training path (`IImageTrainable<T>` is where GS's `TrainOnImageBatch` lives; densification runs inside it)

**Comment by ooples** [comment]
Triage note (PR #1789 sweep): checked against the #1789 branch — **zero** related commits, so that PR neither fixes nor affects this issue. Leaving it open and unlinked; it remains genuine outstanding work.

**Comment by ooples** [comment]
### Re-audit against master (`origin/master` @ `a22d7e78bf` (2026-09-05))

**Why this correction exists.** The `Triage note (PR #1789 sweep)` comment on this issue was produced by checking the **#1789 branch**, which was cut on **2026-07-05**. That sweep ran on **2026-08-14**, the day before #1789 merged. Any fix that reached `master` in that 5.5-week window *outside* #1789 was invisible to it, so those verdicts are false-negative-prone. Re-verified below against master itself.

**Verdict: DONE - recommend close**

Shipped by PR #1838 (merged as `1ca524d4e2`, 2026-07-10). Densification is exposed and schedulable through the facade - 20 citing lines:

- `src/NeuralRadianceFields/Data/IDensificationSchedule.cs:4` - "Densification firing schedule for GaussianSplatting (#1835 excellence goal #2)."
- `src/NeuralRadianceFields/Models/GaussianSplatting.cs:668` - "#1835 densification-schedule getters. Each returns the caller's explicit override"
- `src/Models/Options/GaussianSplattingOptions.cs:13` and `:37` - `EnableDensification = true`, plus the split/prune schedule knobs
- `src/Models/Options/GaussianCompressionOptions.cs:7` - "(#1835 excellence goal #3)"
- `DensificationInterval` is round-tripped in model state (`GaussianSplatting_T_.State.g.cs:16`)

---
*Posted by an automated re-audit of all 45 open issues against master on 2026-09-05. Every figure above was measured on the tree or read from CI run artifacts; none is carried over from a previous comment.*

**Comment by ooples** [comment]
**Triage decision (2026-09-05): staying OPEN.**

Verification went further than code inspection here, and it changed the verdict.

**What is genuinely delivered.** Densification is exposed through the facade and *does* fire from the training loop — not merely configurable:

- `GaussianSplatting.cs:2133` — `if (EnableDensification && schedule.ShouldFire(_trainingStep, lossValue, meanGradNorm)) DensifyAndPrune(gradients);`
- `EnableDensification = true` by default (`GaussianSplattingOptions.cs:37`)
- `IDensificationSchedule` ("#1835 excellence goal #2"), `GaussianCompressionOptions` ("#1835 excellence goal #3"), `SplitChildLearningRateScales` ("#1835 excellence goal #4")
- Split/prune knobs exposed: `OpacityPruneThreshold`, `GradientNormThreshold`, `DensificationInterval`, `DensificationStartIteration/EndIteration`
- Short-run auto-scaling so the paper window `[500, 15000]` does not silently never fire on smaller `MaxIterations`, with explicit values still overriding

**Why it is not closed.** Training with densification enabled **crashes on the step after the first split**:

```
System.ArgumentException : Flat gradient length (224) must match ParameterCount (112).
```

`DensifyAndPrune` resizes the parameter surface but never invalidates the cached parameter layout, so `ParameterCount` keeps describing the previous cloud. Filed as #2089 with the full diagnosis, and fixed there (one line: `InvalidateParameterCountCache()` alongside the existing `MarkSpatialIndexDirty()`), with regression tests that train through the real path.

Existing tests missed this because `GaussianSplattingDensificationTests` drives the internal `RunDensifyAndPruneForTest` hook rather than the training loop, and `ImageTrainingPathTests` sets `EnableDensification = false`.

**Remaining work to close this issue:** confirm #2089's fix lands, then re-verify that a densification-enabled run trains to completion through the facade. The exposure work itself is done and should not be redone.
